creme/creme_api/api/auth/serializers.py

```diff
@@ -0,0 +1,376 @@
+            "last_name",
+            "first_name",
+            "email",
+            "date_joined",
+            "last_login",
+            "is_active",
+            # 'is_staff',
+            "is_superuser",
+            "role",
+            # 'is_team',
+            # 'teammates_set',
+            "time_zone",
+            "theme",
+            # 'settings',
+        ]
+        read_only_fields = [
+            "date_joined",
+            "last_login",
+        ]
+        extra_kwargs = {
+            "first_name": {"required": True},
+            "last_name": {"required": True},
+            "email": {"required": True},
+        }
+
+    def validate(self, attrs):
+        is_superuser = None
+        role_id = None
+
+        if self.instance is not None:
+            is_superuser = self.instance.is_superuser
+            role_id = self.instance.role_id
+
+        has_is_superuser = bool(attrs.get("is_superuser", is_superuser))
+        has_role = bool(attrs.get("role", role_id))
+        if not (has_is_superuser ^ has_role):
+            self.fail("is_superuser_xor_role")
+        return attrs
+
+
+class TeamSerializer(serializers.ModelSerializer):
+    teammates = serializers.PrimaryKeyRelatedField(
+        queryset=CremeUser.objects.filter(is_team=False, is_staff=False),
+        many=True,
+        label=_("Teammates"),
+        required=True,
+        source="teammates_set",
+    )
+
+    class Meta:
+        model = CremeUser
+        fields = [
+            "id",
+            "username",
+            "teammates",
+        ]
+        extra_kwargs = {"username": {"label": _("Team name")}}
+
+    def __init__(self, *args, **kwargs):
+        super(TeamSerializer, self).__init__(*args, **kwargs)
+        username_field = self.fields.pop("username")
+        # username_field.source = "username"
+        self.fields["name"] = username_field
+
+    def save(self, **kwargs):
+        kwargs["is_team"] = True
+        team = super().save(**kwargs)
+        return team
+
+
+class DeleteUserSerializer(serializers.ModelSerializer):
+    """
+    Serializer which assigns the fields with type CremeUserForeignKey
+    referencing a given user A to another user B, then deletes A.
+    """
+
+    transfer_to = serializers.PrimaryKeyRelatedField(
+        queryset=CremeUser.objects.none(), required=True
+    )
+
+    class Meta:
+        model = CremeUser
+        fields = ["transfer_to"]
+
+    def __init__(self, instance=None, **kwargs):
+        super().__init__(instance=instance, **kwargs)
+        users = CremeUser.objects.exclude(is_staff=True)
+        if instance is not None:
+            users = users.exclude(pk=instance.pk)
+        self.fields["transfer_to"].queryset = users
+
+    def save(self, **kwargs):
+        CremeUserForeignKey._TRANSFER_TO_USER = self.validated_data["transfer_to"]
+
+        try:
+            self.instance.delete()
+        finally:
+            CremeUserForeignKey._TRANSFER_TO_USER = None
```

Comment:
Beware the view 'creme.creme_config.views.user.UserDeletion' uses a lock to avoid race conditions (eg: a user transfers A to B, & other one transfers B to A).
- use the same lock (the name is in a class attribute)
- I presume we should perform the deletion in a deletion job (I mean in the creme_config too) ; but I do not know how the REST API would manage that (return immediately "the job has been created" VS wait the job is complete).